consensus/istanbul/core/core.go

```diff
@@ -357,38 +361,41 @@ func (c *core) waitForDesiredRound(r *big.Int) {
 		Round:    new(big.Int).Set(r),
 	}
 	// Perform all of the updates
-	c.setState(StateWaitingForNewRound)
-	c.current.SetDesiredRound(r)
+	c.current.TransitionToWaitingForNewRound(r)
+
 	_, headAuthor := c.backend.GetCurrentHeadBlockAndAuthor()
-	c.valSet.CalcProposer(headAuthor, desiredView.Round.Uint64())
+	c.valSet.CalcProposer(headAuthor, r.Uint64())
 	c.newRoundChangeTimerForView(desiredView)
 
+	// Process Backlog Messages
+	c.processBacklog()
+
 	// Send round change
-	c.sendRoundChange(desiredView.Round)
+	c.sendRoundChange(r)
 }
 
-func (c *core) updateRoundState(view *istanbul.View, validatorSet istanbul.ValidatorSet, roundChange bool) {
+func (c *core) resetRoundState(view *istanbul.View, validatorSet istanbul.ValidatorSet, roundChange bool) {
 	// TODO(Joshua): Include desired round here.
 	if c.current != nil {
 		if roundChange {
-			c.current = newRoundState(view, validatorSet, nil, c.current.PendingRequest(), c.current.PreparedCertificate(), c.current.ParentCommits())
+			c.current = newRoundState(view, validatorSet, c.current.PendingRequest(), c.current.PreparedCertificate(), c.current.ParentCommits())
```

Comment:
Since you added locks everywhere, don't we technically have to have locks here as well to assemble this new roundState? Or are we suffciently protected by the core mutex?

Reply:
actually i'm not actually sure that roundState needs locks tbh, i've added them since they were already there for other methods.

the core engine runs on a single thread, commanded the handleEvents go routine.

So i assume it's safe to skip them here.